src/appservice/MjolnirManager.ts

```diff
@@ -0,0 +1,224 @@
+        );
+        this.mjolnirs.set(await client.getUserId(), managedMjolnir);
+        return managedMjolnir;
+    }
+
+    /**
+     * Gets a mjolnir for the corresponding mxid that is owned by a specific user.
+     * @param mjolnirId The mxid of the mjolnir we are trying to get.
+     * @param ownerId The owner of the mjolnir. We ask for it explicitly to not leak access to another user's mjolnir.
+     * @returns The matching managed mjolnir instance.
+     */
+    public getMjolnir(mjolnirId: string, ownerId: string): ManagedMjolnir|undefined {
+        const mjolnir = this.mjolnirs.get(mjolnirId);
+        if (mjolnir) {
+            if (mjolnir.ownerId !== ownerId) {
+                throw new Error(`${mjolnirId} is owned by a different user to ${ownerId}`);
+            } else {
+                return mjolnir;
+            }
+        } else {
+            return undefined;
+        }
+    }
+
+    /**
+     * Find all of the mjolnirs that are owned by this specific user.
+     * @param ownerId An owner of multiple mjolnirs.
+     * @returns Any mjolnirs that they own.
+     */
+    public getOwnedMjolnirs(ownerId: string): ManagedMjolnir[] {
+        // TODO we need to use the database for this but also provide the utility
+        // for going from a MjolnirRecord to a ManagedMjolnir.
+        // https://github.com/matrix-org/mjolnir/issues/409
+        return [...this.mjolnirs.values()].filter(mjolnir => mjolnir.ownerId !== ownerId);
+    }
+
+    public onEvent(request: Request<WeakEvent>, context: BridgeContext) {
+        // TODO We need a way to map a room id (that the event is from) to a set of managed mjolnirs that should be informed.
+        // https://github.com/matrix-org/mjolnir/issues/412
+        [...this.mjolnirs.values()].forEach((mj: ManagedMjolnir) => mj.onEvent(request));
+    }
+
+    /**
+     * provision a new mjolnir for a matrix user.
+     * @param requestingUserId The mxid of the user we are creating a mjolnir for.
+     * @returns The matrix id of the new mjolnir and its management room.
+     */
+    public async provisionNewMjolnir(requestingUserId: string): Promise<[string, string]> {
+        const access = this.accessControl.getUserAccess(requestingUserId);
+        if (access.outcome !== Access.Allowed) {
+            throw new Error(`${requestingUserId} tried to provision a mjolnir when they do not have access ${access.outcome} ${access.rule?.reason ?? 'no reason specified'}`);
+        }
+        const provisionedMjolnirs = await this.dataStore.lookupByOwner(requestingUserId);
+        if (provisionedMjolnirs.length === 0) {
+            const mjolnirLocalPart = `mjolnir_${randomUUID()}`;
+            const [mjolnirUserId, mjolnirClient] = await this.makeMatrixClient(mjolnirLocalPart);
+
+            const managementRoomId = await mjolnirClient.createRoom({
+                preset: 'private_chat',
+                invite: [requestingUserId],
+                name: `${requestingUserId}'s mjolnir`
```

Comment:
So, if I understand correctly, we have one Mjölnir and management room per user? I thought that it was one Mjölnir and management room per protected room. Having one Mjölnir per user scares me a bit:

1. Several users can have a Mjölnir for the same room, which strikes me as suboptimal.
2. If user A creates a room and a Mjölnir, then they cannot give away this Mjölnir to user B when they decide that they're tired of being the moderator for the room?
3. I suspect that this will complicate life when we proceed with decentralizing abuse reports, although I cannot articulate exactly why.

Am I missing something?

Reply:
Maybe it would be wise to call them creators rather than owners

Reply:
So why do we need to know the creator of a room? Is it just to be able to revoke privileges if they misbehave?

Reply:
We need to know who has requested the mjolnir to be provisoined. All it is for is so that we can control who can create them and monitor how many are being created (imagine any server with public registration also allowing their users to create mjolnirs). If the creator becomes inactive then the mjolnir will keep working.

Reply:
What I mean is that we can have the following:

1. User A is in good standing.
2. User A creates a Mjölnir.
3. User A steps down, letting user B be the de facto owner of this Mjölnir.
4. User A becomes an obnoxious troll and gets kicked out from the server.
5. We remove A's authorization to generate Mjölniren (I think I'm going to keep that plural :) ).
6. Any Mjölnir of A gets closed.
7. B loses a Mjölnir.

The only reason I can think of for indexing Mjölniren by creator is to do 6, no?

Reply:
I don't want to get (too much) in the way of early testing, but I believe that changing this behavior so that B doesn't lose a Mjölnir should be addressed soon-ish. Could you file an issue?